frost-secp256k1-tr/src/lib.rs

```diff
@@ -0,0 +1,755 @@
+    Scalar::from(&sp)
+}
+
+/// Create a BIP340 compliant tagged hash
+fn tagged_hash(tag: &str) -> Sha256 {
+    let mut hasher = Sha256::new();
+    let mut tag_hasher = Sha256::new();
+    tag_hasher.update(tag.as_bytes());
+    let tag_hash = tag_hasher.finalize();
+    hasher.update(tag_hash);
+    hasher.update(tag_hash);
+    hasher
+}
+
+/// Create a BIP341 compliant taproot tweak
+fn tweak<T: AsRef<[u8]>>(
+    public_key: &<<Secp256K1Sha256 as Ciphersuite>::Group as Group>::Element,
+    merkle_root: Option<T>,
+) -> Scalar {
+    match merkle_root {
+        None => Secp256K1ScalarField::zero(),
+        Some(root) => {
+            let mut hasher = tagged_hash("TapTweak");
+            hasher.update(public_key.to_affine().x());
+            hasher.update(root.as_ref());
+            hasher_to_scalar(hasher)
+        }
+    }
+}
+
+/// Create a BIP341 compliant tweaked public key
+fn tweaked_public_key<T: AsRef<[u8]>>(
+    public_key: &VerifyingKey,
+    merkle_root: Option<T>,
+) -> <<Secp256K1Sha256 as Ciphersuite>::Group as Group>::Element {
+    let mut pk = public_key.to_element();
+    if pk.to_affine().y_is_odd().into() {
+        pk = -pk;
+    }
+    ProjectivePoint::GENERATOR * tweak(&pk, merkle_root) + pk
+}
+
+/// The message target which the group's signature should commit to. Includes
+/// a message byte vector, and a set of ciphersuite-specific parameters.
+pub type SigningTarget = frost_core::SigningTarget<S>;
+
+/// The ciphersuite-specific signing parameters which are fed into
+/// signing code to ensure correctly compliant signatures are computed.
+#[derive(Debug, Clone, Eq, PartialEq, Default)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub struct SigningParameters {
+    /// The tapscript merkle tree root which must be committed to and agreed upon
+    /// in advance by all participants in the signing round.
+    ///
+    /// If set to `None` (the default), then no taproot tweak will be committed to in the signature.
+    /// Best practice suggested by BIP341 is to commit to an empty merkle root in cases
+    /// where no tapscript tweak is needed, i.e. by supplying `&[0; u8]` as the merkle root.
+    /// This prevents hiding of taproot commitments inside a linearly aggregated key.
+    ///
+    /// However, for FROST, this is not strictly required as the group key cannot be
+    /// poisoned as long as the DKG procedure is conducted correctly.
+    /// Thus, the [`Default`] trait implementation of taproot `SigningParameters`
+    /// sets `tapscript_merkle_root` to `None`.
+    ///
+    /// If 3rd party observers outside the FROST group must be able to verify there
+    /// is no hidden script-spending path embedded in the FROST group's taproot output key,
+    /// then you should set `tapscript_merkle_root` to `Some(vec![])`, which proves
+    /// the tapscript commitment for the tweaked output key is unspendable.
+    pub tapscript_merkle_root: Option<Vec<u8>>,
```

Comment:
@conduition It was a while ago that I wrote my own code but I went back and I see this is where the tweak can be provided if this is set to an empty vector. If this is set to None for a key-spend then that is where the problem lies.

In my own code I set this to None when providing signatures for a script-path spend. I don't believe there is any issue with that. I pass an empty vector when a key-spend is done without a MAST hash which then includes the necessary tweak.

Provided that consumers of the library are passing an empty vector when doing a MAST-free key-spend, I don't see there being an issue. However there might be a problem if this is used incorrectly and None is accidentally used. An additional tweak as part of the DKG process might still be a good idea to protect against misuse.

Maybe the API can be improved with factory methods such as `SigningParameters::keySpend` and `SigningParameters::scriptSpend` so it's clear when a key is used for key-path or script-path signatures and where key-path keys are always tweaked.

I've written a Dart wrapper that uses these high-level Dart constructors:

```dart
  /// Used for key-spend signatures. This will always tweak the internal key. If
  /// there is no MAST root hash provided, the tweak will be done without a
  /// MAST hash for Taproot programs that only have a key-path.
  SignDetails.keySpend({
    required Uint8List message,
    Uint8List? mastHash,
  }) : this(message: message, mastHash: mastHash ?? Uint8List(0));

  /// Used for script-spend signatures. This does not tweak the key allowing for
  /// direct signing with a key specified in a Tapscript.
  SignDetails.scriptSpend({ required Uint8List message })
    : this(message: message, mastHash: null);

```

Reply:
> In my own code I set this to None when providing signatures for a script-path spend. I don't believe there is any issue with that.

Yes, you're correct. In fact I have some clients who are doing exactly this (script-path spending with FROST signatures in the script witness)

> Provided that consumers of the library are passing an empty vector when doing a MAST-free key-spend, I don't see there being an issue.

That's exactly the issue: We want security _by default_. Key-spending is the default for P2TR, so we can't allow key-spending users of this crate to footgun themselves just because they didn't know to explicitly configure their verifying key with the right arcane parameter of `Some(vec![])`.

> Maybe the API can be improved with factory methods such as SigningParameters::keySpend and SigningParameters::scriptSpend so it's clear when a key is used for key-path or script-path signatures and where key-path keys are always tweaked.

Technically you can use tweaked or untweaked BIP340 signatures for either key-spending or script-spending, there's no hard relationship there. 

- Maybe I'd want to tweak a pubkey inside the MAST tapscript leaf to commit to some custom data. Even if bitcoin nodes don't care or validate that tweak, it's still a useful option for me in off-chain protocols.
- Likewise, it's perfectly fine for a single-signer to use an untweaked plain public key as a P2TR output and then key-spend from that.

The only dangerous combination is using an untweaked _linearly aggregated_ pubkey in a P2TR output script, _if_ the keys which compose that aggregated pubkey come from untrusted sources.

Reply:
Yes, I think it's fair to tweak the DKG output to avoid misuse. I included named constructors in my code to avoid misuse but it's not difficult to add a tweak during the DKG process.